src/urllib3/response.py

```diff
@@ -1020,10 +1022,13 @@ def stream(
             yield from self.read_chunked(amt, decode_content=decode_content)
         else:
             while not is_fp_closed(self._fp) or len(self._decoded_buffer) > 0:
-                data = self.read(amt=amt, decode_content=decode_content)
-
-                if data:
-                    yield data
+                if amt is None:
+                    data = self.read1(amt=amt, decode_content=decode_content)
+                else:
+                    data = self.read(amt=amt, decode_content=decode_content)
+                if not data:
+                    break
```

Comment:
> When the loop is broken here after `read1` was used, `self._fp` remains not closed even though all data has been read. 

Hadn't realised the semantics around what gets closed and when, have done a bit more reading now.  Have had a bit of a refactor of the change.

> Closing it may be a better way to break the loop.  What do you think about checking `self.length_remaining` after calling `read1` and, if it is 0 and `not self.closed`, calling `self._fp.close()`?

It feels as though it could raise an `IncompleteRead` if it finishes and `length_remaining` indicates there's missing data (as `_raw_read` does), any preferences on this / other changes

Reply:
> It feels as though it could raise an `IncompleteRead` if it finishes and `length_remaining` indicates there's missing data (as `_raw_read` does), any preferences on this / other changes

Good point!
Looking at [CPython's code](https://github.com/python/cpython/blob/2d76be251d0aee89f76e6fa5a63fa1ad3f2b76cf/Lib/http/client.py#L657-L658), it's possible to get in the situation at least if `self._fp` is closed between `read1` calls. However, you have to respect `self.enforce_content_length` as  `_raw_read` does.